src/charm.py

```diff
@@ -275,8 +304,84 @@ def _reconcile(self) -> None:
+        )
         self._set_status_gateway_address(client, gateway_resource_information)
 
+    def _update_dns_record_relation(
+        self,
+        external_hostname: str,
+        client: Client,
+        gateway_resource_information: GatewayResourceInformation,
+    ) -> None:
+        """Update the DNS record relation with the external hostname and gateway address.
+
+        Args:
+            external_hostname: The external hostname to be used in the DNS record.
+            client: Lightkube client.
+            gateway_resource_information: Information needed to create the gateway resource.
+        """
+        relation = self.model.get_relation(self.dns_record_requirer.relation_name)
+        if not relation or not external_hostname:
+            return
+        resource_manager = GatewayResourceManager(
+            labels=self._labels,
+            client=client,
+        )
+        if not resource_manager.current_gateway_resource():
+            logger.warning(
+                "No gateway resource found, cannot update DNS record for %s",
+                external_hostname,
+            )
+            return
+        gateway_address = resource_manager.gateway_address(
+            gateway_resource_information.gateway_name
+        )
+        if not gateway_address:
+            logger.warning(
+                "No gateway address found for %s, cannot update DNS record",
+                external_hostname,
+            )
+            return
+        host_label, domain = self._split_hostname(external_hostname)
+        entry = RequirerEntry(
+            domain=domain,
+            host_label=host_label,
+            ttl=600,
+            record_class=RecordClass.IN,
+            record_type=RecordType.A,
+            record_data=gateway_address,
+            uuid=uuid.uuid5(UUID_NAMESPACE, str(external_hostname) + str(gateway_address)),
+        )
+        dns_record_requirer_data = DNSRecordRequirerData(dns_entries=[entry])
+        self.dns_record_requirer.update_relation_data(relation, dns_record_requirer_data)
+
+    def _split_hostname(self, hostname: str) -> Tuple[str, str]:
+        """Split external_hostname into host_label and domain.
+
+        - host_label: the subdomain or specific host (e.g. 'www')
+        - domain: the root domain (e.g. 'example.com')
+
+        If the hostname is a root domain (like 'example.com'), the host_label is '@'.
+
+        Raises:
+            ValueError: if the hostname is invalid.
+        """
+        labels = hostname.split(".")
+
+        if len(labels) < 2:
+            raise ValueError(f"Hostname must contain at least one dot: {hostname}")
+
+        if len(labels) == 2:
+            # Apex/root domain (e.g., 'example.com')
+            return "@", hostname
+
+        # More than 2 labels: e.g., 'www.example.com' → ('www', 'example.com')
+        host_label = labels[0]
+        domain = ".".join(labels[1:])
+        return host_label, domain
+
```

Comment:
@gregory-schiano 
I think that now is the time to discuss this splitting function.
I see two issues here:
1) The registries can give away TLDs like `canonical.co.uk` that wouldn't work here and it is impossible to know those cases without a list of these occurences.
2) `("@", "www.example.com")` is not exactly the same as `("www", "example.com")` as we could only have ownership over `www.example.com` on the `bind-operator` side.

I feel that this is an information that may need to be conveyed to the ingress